### ISSUE TITLE
Switch case assignments failed to narrow type without extraneous default branch

### ISSUE THREAD

**Issue by patrickroberts** [Microsoft/TypeScript#26836]
**TypeScript Version:**  3.1.0-dev.20180901

**Search Terms:** switch case assignment narrow type default

**Code**

```ts
/// <reference lib="es2015" />

type NodeType = 'a'

interface Node {
  type: NodeType
}

const cache = new WeakMap<Node, string>()

function test(node: Node): string {
  let result = cache.get(node)

  if (!result) {
    switch (node.type) {
      case 'a':
        result = '1'
        break
    }

    cache.set(node, result)
  }

  return result
}
```

**Expected behavior:** No errors when compiling with `--strict` since `result`'s type has been narrowed to exclude `undefined` based on the possible values of `node.type`.

**Actual behavior:**

```
test.ts:21:21 - error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

21     cache.set(node, result)
                       ~~~~~~

test.ts:24:3 - error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.

24   return result
     ~~~~~~~~~~~~~
```

**Playground Link:** 
[TypeScript Playground demo](https://www.typescriptlang.org/play/#src=%2F%2F%2F%20%3Creference%20lib%3D%22es2015%22%20%2F%3E%0D%0A%0D%0Atype%20NodeType%20%3D%20'a'%0D%0A%0D%0Ainterface%20Node%20%7B%0D%0A%20%20type%3A%20NodeType%0D%0A%7D%0D%0A%0D%0Aconst%20cache%20%3D%20new%20WeakMap%3CNode%2C%20string%3E()%0D%0A%0D%0Afunction%20test(node%3A%20Node)%3A%20string%20%7B%0D%0A%20%20let%20result%20%3D%20cache.get(node)%0D%0A%0D%0A%20%20if%20(!result)%20%7B%0D%0A%20%20%20%20switch%20(node.type)%20%7B%0D%0A%20%20%20%20%20%20case%20'a'%3A%0D%0A%20%20%20%20%20%20%20%20result%20%3D%20'1'%0D%0A%20%20%20%20%20%20%20%20break%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20cache.set(node%2C%20result)%0D%0A%20%20%7D%0D%0A%0D%0A%20%20return%20result%0D%0A%7D)

**Related Issues:** 
https://github.com/Microsoft/TypeScript/issues/13441

**Comment by ahejlsberg** [comment]
This is a design limitation. The control flow analyzer doesn't track dependencies between distinct variables so it is unable to conclude that `result` is definitely assigned following the `switch` statement. All it knows is that, ignoring effects of narrowing `node.type`, there is a possible code path in which the `switch` statement doesn't assign to `result`, and therefore it assumes that `result` might still be `undefined`.

**Comment by patrickroberts** [comment]
@ahejlsberg Thanks for explaining the cause of this issue. Do you have any recommendations for an alternative control flow that avoids this limitation without adding a branch that will never get called?

**Comment by patrickroberts** [comment]
I came up with a work-around by splitting the function into its basic form and a memoization helper. This allows the analyzer to track dependencies between `node.type` and the return value, rather than two distinct variables.

```ts
/// <reference lib="es2015" />

type NodeType = 'a'

interface Node {
  type: NodeType
}

function memoize<I extends object, O> (fn: (input: I) => O) {
  const cache = new WeakMap<I, O>()

  return (input: I) => {
    let output: O | undefined = cache.get(input)

    if (!output) {
      output = fn(input)
      cache.set(input, output)
    }

    return output
  }
}

const test = memoize((node: Node): string => {
  switch (node.type) {
    case 'a':
      return '1'
      break
  }
})
```

[TypeScript Playground demo](https://www.typescriptlang.org/play/#src=%2F%2F%2F%20%3Creference%20lib%3D%22es2015%22%20%2F%3E%0D%0A%0D%0Atype%20NodeType%20%3D%20'a'%0D%0A%0D%0Ainterface%20Node%20%7B%0D%0A%20%20type%3A%20NodeType%0D%0A%7D%0D%0A%0D%0Afunction%20memoize%3CI%20extends%20object%2C%20O%3E%20(fn%3A%20(input%3A%20I)%20%3D%3E%20O)%20%7B%0D%0A%20%20const%20cache%20%3D%20new%20WeakMap%3CI%2C%20O%3E()%0D%0A%0D%0A%20%20return%20(input%3A%20I)%20%3D%3E%20%7B%0D%0A%20%20%20%20let%20output%3A%20O%20%7C%20undefined%20%3D%20cache.get(input)%0D%0A%0D%0A%20%20%20%20if%20(!output)%20%7B%0D%0A%20%20%20%20%20%20output%20%3D%20fn(input)%0D%0A%20%20%20%20%20%20cache.set(input%2C%20output)%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20return%20output%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20test%20%3D%20memoize((node%3A%20Node)%3A%20string%20%3D%3E%20%7B%0D%0A%20%20switch%20(node.type)%20%7B%0D%0A%20%20%20%20case%20'a'%3A%0D%0A%20%20%20%20%20%20return%20'1'%0D%0A%20%20%20%20%20%20break%0D%0A%20%20%7D%0D%0A%7D))
